### PR TITLE
feat: add support for application command permissions v2

### PR DESCRIPTION
**Pull request by suneettipirneni**
**Please describe the changes this PR makes and why it should be merged:**

Credit to @vitorlops for making the basis for this PR.

**If applicable, please reference Discord API Docs PRs or commits that influenced this PR:**

- https://github.com/discord/discord-api-docs/pull/4830
- https://github.com/discord/discord-api-docs/pull/4836
- https://github.com/discord/discord-api-docs/pull/4833
- https://github.com/discord/discord-api-docs/pull/4845
- https://github.com/discord/discord-api-docs/pull/4861

### CODE REVIEW

File: deno/payloads/v10/_interactions/_applicationCommands/permissions.ts [hunk 1]
```diff
@@ -46,4 +46,26 @@ export interface APIApplicationCommandPermission {
...
+/**
+ * https://discord.com/developers/docs/interactions/application-commands#application-command-permissions-object-application-command-permissions-constants
+ */
+export const APIApplicationCommandPermissionsConstant = {
+	Everyone(guildId: string | bigint): Snowflake {
+		if (typeof guildId === 'bigint') {
+			return `${guildId}`;
+		}
+
+		return guildId;
```
Comment (vladfrangu) [inline deno/payloads/v10/_interactions/_applicationCommands/permissions.ts:61]:
```suggestion
		return String(guildId);
```

File: deno/payloads/v10/_interactions/applicationCommands.ts [hunk 2]
```diff
@@ -70,6 +70,14 @@ export interface APIApplicationCommand {
...
+	/**
+	 * Set of permissions represented as a bitset
+	 */
+	default_member_permissions?: Permissions;
+	/**
+	 * Indicates whether the command is available in DMs with the app, only for globally-scoped commands. By default, commands are visible
+	 */
+	dm_permission?: boolean;
 	/**
 	 * Whether the command is enabled by default when the app is added to a guild
```
Comment (vladfrangu) [inline deno/payloads/v10/_interactions/applicationCommands.ts:82]: Should deprecate this

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Comment by vladfrangu** [comment]
Should also track https://github.com/discord/discord-api-docs/pull/4836 and https://github.com/discord/discord-api-docs/pull/4833

File: payloads/v9/auditLog.ts [hunk 3]
```diff
@@ -517,6 +519,16 @@ export type APIAuditLogChangeKeyPermissions = AuditLogChangeData<'permissions',
  */
 export type APIAuditLogChangeKeyColor = AuditLogChangeData<'color', number>;
 
+/**
+ * Returned when application command permissions are changed
+ */
+export type APIAuditLogChangeKeyCommandId = AuditLogChangeData<'command_id', string>;
```
Comment (ckohen) [inline payloads/v9/auditLog.ts:525]: This one doesn't appear to actually exist, waiting for confirmation on it

File: payloads/v9/auditLog.ts [hunk 4]
```diff
@@ -517,6 +519,16 @@ export type APIAuditLogChangeKeyPermissions = AuditLogChangeData<'permissions',
...
+
+/**
+ * Returned when permissions for a command were updated
+ */
+export type APIAuditLogChangeKeySnowflake = AuditLogChangeData<'snowflake', Snowflake>;
+
```
Comment (ckohen) [inline payloads/v9/auditLog.ts:531]: this one is an object currently, again waiting on confirmation

File: deno/payloads/v10/_interactions/_applicationCommands/permissions.ts [hunk 5]
```diff
@@ -27,11 +27,11 @@ export interface APIGuildApplicationCommandPermissions {
  */
 export interface APIApplicationCommandPermission {
 	/**
-	 * The id of the role or user
+	 * The id of the role, user or channel
```
Comment (vitorlops) [inline deno/payloads/v10/_interactions/_applicationCommands/permissions.ts:30]:
```suggestion
	 * The id of the role, user or channel. Can also be a permission constant
```

File: deno/payloads/v10/_interactions/applicationCommands.ts [hunk 6]
```diff
@@ -70,6 +70,14 @@ export interface APIApplicationCommand {
 	 * The parameters for the `CHAT_INPUT` command, max 25
 	 */
 	options?: APIApplicationCommandOption[];
+	/**
+	 * Set of permissions represented as a bitset
+	 */
+	default_member_permissions?: Permissions;
```
Comment (vitorlops) [inline deno/payloads/v10/_interactions/applicationCommands.ts:76]:
```suggestion
	default_member_permissions?: Permissions | null;
```
https://github.com/discord/discord-api-docs/pull/4830/commits/0dc91010bb37ed059a03a6ec52f6dc110ff3b684

File: deno/payloads/v10/oauth2.ts [hunk 7]
```diff
@@ -111,4 +111,10 @@ export enum OAuth2Scopes {
 	 * See https://discord.com/developers/docs/interactions/application-commands
 	 */
 	ApplicationsCommandsUpdate = 'applications.commands.update',
+	/**
+	 * Allows your app to update permissions for its commands using a Bearer token - client credentials grant only
```
[the remark this answers is not in the document]
Comment (vladfrangu) [reply deno/payloads/v10/oauth2.ts:115]: We usually copy paste doc strings directly from Discord so.. some strings may look out of place

File: deno/payloads/v10/_interactions/_applicationCommands/permissions.ts (hunk 5 above)
Comment (vladfrangu) [reply deno/payloads/v10/_interactions/_applicationCommands/permissions.ts:30]: Is this documented on Discord's side? If not it should be

Comment (vitorlops) [reply deno/payloads/v10/_interactions/_applicationCommands/permissions.ts:30]: It is yes
image

**Comment by vitorlops** [comment]
About https://github.com/discordjs/discord-api-types/pull/415#issuecomment-1111959991, more one to track https://github.com/discord/discord-api-docs/pull/4845

File: deno/payloads/v10/_interactions/applicationCommands.ts [hunk 8]
```diff
@@ -70,6 +70,14 @@ export interface APIApplicationCommand {
...
+	/**
+	 * Indicates whether the command is available in DMs with the app, only for globally-scoped commands. By default, commands are visible
+	 */
+	dm_permission?: boolean;
```
[the remark this answers is not in the document]
Comment (ckohen) [reply deno/payloads/v10/_interactions/applicationCommands.ts:80]: This is changing, likely before this PR lands

File: rest/v10/interactions.ts [hunk 9]
```diff
@@ -71,7 +71,9 @@ export type RESTPostAPIApplicationCommandsResult = APIApplicationCommand;
 /**
  * https://discord.com/developers/docs/interactions/application-commands#edit-global-application-command
  */
-export type RESTPatchAPIApplicationCommandJSONBody = StrictPartial<RESTPostAPIApplicationCommandsJSONBody>;
+export type RESTPatchAPIApplicationCommandJSONBody = StrictPartial<
+	Omit<RESTPostAPIApplicationCommandsJSONBody, 'dm_permission'>
+>;
```
Comment (ckohen) [inline rest/v10/interactions.ts:76]: Did the wrong one here, should be all the guild endpoints - post, patch, and put

File: deno/payloads/v10/_interactions/_applicationCommands/permissions.ts (hunk 1 above)
Comment (vladfrangu) [reply deno/payloads/v10/_interactions/_applicationCommands/permissions.ts:61]: This was not resolved

File: deno/payloads/v10/auditLog.ts [hunk 10]
```diff
@@ -27,6 +28,12 @@ import type { Snowflake } from '../../globals.ts';
  * https://discord.com/developers/docs/resources/audit-log#audit-log-object-audit-log-structure
  */
 export interface APIAuditLog {
+	/**
+	 * List of application commands found in the audit log
+	 *
+	 * See https://discord.com/developers/docs/interactions/application-commands#application-command-object
+	 */
+	application_commands: APIApplicationCommand[];
```
Comment (vladfrangu) [inline deno/payloads/v10/auditLog.ts:36]: Is this field always present (even for audit log entries that do not affect a command) or can it be undefined too?

File: deno/payloads/v10/auditLog.ts [hunk 11]
```diff
@@ -517,6 +526,16 @@ export type APIAuditLogChangeKeyPermissions = AuditLogChangeData<'permissions',
...
 
+/**
+ * Returned when application command permissions are changed
+ */
+export type APIAuditLogChangeKeyCommandId = AuditLogChangeData<'command_id', string>;
+
+/**
+ * Returned when permissions for a command were updated
+ */
+export type APIAuditLogChangeKeySnowflake = AuditLogChangeData<'snowflake', Snowflake>;
```
Comment (vladfrangu) [inline deno/payloads/v10/auditLog.ts:537]: These two are wrong, one of the possibilities is now the `change.key` can be a Snowflake and have a value that for all intents and purposes we could only strictly type as unknown... (for this one PR the value is APICommandPermissionData or w/e)

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: deno/payloads/v10/auditLog.ts [hunk 12]
```diff
@@ -517,6 +526,11 @@ export type APIAuditLogChangeKeyPermissions = AuditLogChangeData<'permissions',
  */
 export type APIAuditLogChangeKeyColor = AuditLogChangeData<'color', number>;
 
+/**
+ * Returned when permissions for a command were updated
+ */
+export type APIAuditLogChangeKeySnowflake = AuditLogChangeData<'snowflake', unknown>;
```
Comment (vladfrangu) [inline deno/payloads/v10/auditLog.ts:532]: The key is a Snowflake. It's not 'snowflake'

File: deno/payloads/v9/_interactions/_applicationCommands/permissions.ts [hunk 13]
```diff
@@ -46,4 +46,19 @@ export interface APIApplicationCommandPermission {
...
+ */
+export const APIApplicationCommandPermissionsConstant = {
+	Everyone: (guildId: string | bigint): Snowflake => String(guildId),
+	AllChannels(guildId: string | bigint): Snowflake {
+		if (typeof guildId === 'string') {
+			return `${BigInt(guildId) - 1n}`;
+		}
+
+		return `${guildId - 1n}`;
+	},
```
Comment (vladfrangu) [inline deno/payloads/v9/_interactions/_applicationCommands/permissions.ts:63]:
```suggestion
	AllChannels: (guildId: string | bigint): Snowflake => String(BigInt(guildId) - 1n) 
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: deno/payloads/v10/auditLog.ts [hunk 14]
```diff
@@ -27,6 +28,12 @@ import type { Snowflake } from '../../globals.ts';
  * https://discord.com/developers/docs/resources/audit-log#audit-log-object-audit-log-structure
  */
 export interface APIAuditLog {
+	/**
+	 * List of application commands found in the audit log
+	 *
+	 * See https://discord.com/developers/docs/interactions/application-commands#application-command-object
+	 */
+	application_commands?: APIApplicationCommand[];
```
Comment (vladfrangu) [inline deno/payloads/v10/auditLog.ts:36]: I was wrong, seems that this is always returned, only as an empty array 🥴

File: deno/payloads/v10/auditLog.ts [hunk 15]
```diff
@@ -517,6 +526,11 @@ export type APIAuditLogChangeKeyPermissions = AuditLogChangeData<'permissions',
  */
 export type APIAuditLogChangeKeyColor = AuditLogChangeData<'color', number>;
 
+/**
+ * Returned when permissions for a command were updated
```
Comment (vladfrangu) [inline deno/payloads/v10/auditLog.ts:530]:
```suggestion
 * Represents a change where the key is a snowflake.
 * Currently, the only known instance of this is returned when permissions for a command were updated (<insert name of object here>)
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]
final requests i swear

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**